### PR TITLE
[Core][Hotfix] Correcting compute of the residual/displacement norm

### PR DESCRIPTION
**Pull request by loumalouomega**

### CODE REVIEW

**Comment by loumalouomega** [comment]
@KratosMultiphysics/technical-committee

File: kratos/solving_strategies/convergencecriterias/displacement_criteria.h
Comment (RiccardoRossi) [inline kratos/solving_strategies/convergencecriterias/displacement_criteria.h]: i think this does happen. I wouldn't throw an error here...

File: kratos/solving_strategies/convergencecriterias/displacement_criteria.h
Comment (RiccardoRossi) [inline kratos/solving_strategies/convergencecriterias/displacement_criteria.h:348]: i just realized that if the dof_is_fixed than the dx is zero, so it was not harming not to do this check :-(

File: kratos/solving_strategies/convergencecriterias/residual_criteria.h
Comment (RiccardoRossi) [inline kratos/solving_strategies/convergencecriterias/residual_criteria.h:349]: while this one IS important, since the residual will be nonzero even at convergence when the dof is fixed

**Review by RiccardoRossi — COMMENTED** [review]
almost there. I think we shouldn't throw an error if the initial norm is zero, but think at some solution for it...(for example if we find the initial norm to be zero, set it to the norm we just found/1000 or something of the sort

File: kratos/solving_strategies/convergencecriterias/displacement_criteria.h
Comment (loumalouomega) [reply kratos/solving_strategies/convergencecriterias/displacement_criteria.h:348]: In the displacement criteria to ponderate the number of dofs is considered, then if affects

File: kratos/solving_strategies/convergencecriterias/displacement_criteria.h
Comment (loumalouomega) [reply kratos/solving_strategies/convergencecriterias/displacement_criteria.h]: Ok

**Comment by loumalouomega** [comment]
is this OK?

**Comment by RiccardoRossi** [comment]
to me this looks fine (good job Vicente!). Does anyone have anything against merging this? (i am afraid it might change the tolerance in some tests)

**Comment by loumalouomega** [comment]
Checked, all test run with the last modification

**Comment by loumalouomega** [comment]
Changes were applied @RiccardoRossi

File: kratos/solving_strategies/convergencecriterias/residual_criteria.h
Comment (RiccardoRossi) [inline kratos/solving_strategies/convergencecriterias/residual_criteria.h]: at this point you can cleanup the commented lines

File: kratos/solving_strategies/convergencecriterias/displacement_criteria.h
Comment (RiccardoRossi) [inline kratos/solving_strategies/convergencecriterias/displacement_criteria.h]: why did u change the echo level here?

**Review by RiccardoRossi — DISMISSED** [review]
this looks fine to me, with two kludges:

1 - @loumalouomega did u also run the mpi tests?
2 - can anyone also take a look at this? given the potential impact of the change i would like to have 2 approvals (i mean another one apart for me)

File: kratos/solving_strategies/convergencecriterias/displacement_criteria.h
Comment (loumalouomega) [reply kratos/solving_strategies/convergencecriterias/displacement_criteria.h]: > 0 is equal to >= 1

File: kratos/solving_strategies/convergencecriterias/residual_criteria.h
Comment (loumalouomega) [reply kratos/solving_strategies/convergencecriterias/residual_criteria.h]: Ok

**Comment by loumalouomega** [comment]
The MPI tests are no working right now apparently

**Comment by loumalouomega** [comment]
Can anyone re-approve?

**Comment by loumalouomega** [comment]
@RiccardoRossi ping

**Comment by loumalouomega** [comment]
Ping @RiccardoRossi

**Comment by loumalouomega** [comment]
Ping

File: kratos/solving_strategies/convergencecriterias/displacement_criteria.h
Comment (RiccardoRossi) [inline kratos/solving_strategies/convergencecriterias/displacement_criteria.h]: i think this is not needed

File: kratos/solving_strategies/convergencecriterias/displacement_criteria.h
Comment (RiccardoRossi) [inline kratos/solving_strategies/convergencecriterias/displacement_criteria.h]: this is not a ratio ... this is the norm at the beginning of the iterations

File: kratos/solving_strategies/convergencecriterias/displacement_criteria.h
Comment (RiccardoRossi) [reply kratos/solving_strategies/convergencecriterias/displacement_criteria.h:348]: how is this different from the previous funciton?

->because previously it wasn't checking the DoF fixed, this is precisely the motivation of this PR

File: kratos/solving_strategies/convergencecriterias/residual_criteria.h
Comment (RiccardoRossi) [inline kratos/solving_strategies/convergencecriterias/residual_criteria.h]: comment is wrong

File: kratos/solving_strategies/convergencecriterias/displacement_criteria.h
Comment (RiccardoRossi) [reply kratos/solving_strategies/convergencecriterias/displacement_criteria.h]: this if and else is definitely not needed

File: kratos/solving_strategies/convergencecriterias/displacement_criteria.h
Comment (RiccardoRossi) [reply kratos/solving_strategies/convergencecriterias/displacement_criteria.h]: no ... 0 < 1

File: kratos/solving_strategies/convergencecriterias/displacement_criteria.h
Comment (loumalouomega) [reply kratos/solving_strategies/convergencecriterias/displacement_criteria.h]: Why?, this was already in the code, I just changed the 0 for epsilon

File: kratos/solving_strategies/convergencecriterias/residual_criteria.h
Comment (loumalouomega) [reply kratos/solving_strategies/convergencecriterias/residual_criteria.h]: Ok

**Comment by loumalouomega** [comment]
Done

**Comment by loumalouomega** [comment]
Ping

**Comment by loumalouomega** [comment]
Ping @RiccardoRossi

File: kratos/solving_strategies/convergencecriterias/residual_criteria.h
Comment (RiccardoRossi) [inline kratos/solving_strategies/convergencecriterias/residual_criteria.h]: take out the commented code

File: kratos/solving_strategies/convergencecriterias/residual_criteria.h
Comment (RiccardoRossi) [inline kratos/solving_strategies/convergencecriterias/residual_criteria.h]: remove

File: kratos/solving_strategies/convergencecriterias/displacement_criteria.h
Comment (RiccardoRossi) [inline kratos/solving_strategies/convergencecriterias/displacement_criteria.h]: rename this tolerance to something like "approx_zero_tolerance" or something of the sort

File: kratos/solving_strategies/convergencecriterias/displacement_criteria.h
Comment (RiccardoRossi) [reply kratos/solving_strategies/convergencecriterias/displacement_criteria.h]: reference_norm will not be zero, hence this will be at worst a 1

**Review by RiccardoRossi — APPROVED** [review]
approving (provided that tests pass)

File: kratos/solving_strategies/convergencecriterias/displacement_criteria.h
Comment (philbucher) [inline kratos/solving_strategies/convergencecriterias/displacement_criteria.h:135]: @loumalouomega this does not compile with trilinos!

File: kratos/solving_strategies/convergencecriterias/displacement_criteria.h
Comment (philbucher) [inline kratos/solving_strategies/convergencecriterias/displacement_criteria.h:350]: neither does this

Not sure if it is ok to do TSparseSpace::GetValue though bcs the entry could also be non-local I think (then in MPI it would throw)

**Review by philbucher — COMMENTED** [review]
the modifications do not compile in MPI!

Can you please check?

I am not sure how to fix it otherwise I would have done it myself

**Comment by loumalouomega** [comment]
Can you solve it?, I am this week off.

**Comment by philbucher** [comment]
ok

File: kratos/solving_strategies/convergencecriterias/displacement_criteria.h
Comment (ddiezrod) [inline kratos/solving_strategies/convergencecriterias/displacement_criteria.h:138]: @RiccardoRossi @loumalouomega I know I am arriving a little late here, but this used the two norm before, so it returned the square root of this number.  Is this change intentional? I do not really know which one should be used. (@jrubiogonzalez )

Comment (loumalouomega) [reply kratos/solving_strategies/convergencecriterias/displacement_criteria.h:138]: Hi @ddiezrod you are right, @marandra mentioned me about it, and I forgot to revert it. One sec

**Pull request merged**
